File: app/src/main/kotlin/com/openclaw/ghostcrab/data/impl/InstalledSkillRepositoryImpl.kt
package com.openclaw.ghostcrab.data.impl

import com.openclaw.ghostcrab.data.ws.GatewayWsClient
import com.openclaw.ghostcrab.domain.model.InstalledSkill
import com.openclaw.ghostcrab.domain.model.SkillInstallError
import com.openclaw.ghostcrab.domain.model.SkillInstallProgress
import com.openclaw.ghostcrab.domain.model.SkillSource
import com.openclaw.ghostcrab.domain.repository.InstalledSkillRepository
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.channelFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlinx.serialization.json.put

/**
 * WS-backed [InstalledSkillRepository]. Callers provide a factory so each op obtains
 * the live session for the currently connected gateway — mirrors the lazy lookup used
 * by [GatewayConnectionManagerImpl].
 *
 * @param wsFactory Suspending factory that returns the active [GatewayWsClient].
 */
@Suppress("TooGenericExceptionCaught")
class InstalledSkillRepositoryImpl(
    private val wsFactory: suspend () -> GatewayWsClient,
) : InstalledSkillRepository {

    private val installedFlow = MutableStateFlow<List<InstalledSkill>>(emptyList())

    /**
     * Hot [Flow] of the current installed-skill list. Starts empty; updated on
     * successful install/uninstall and after [refreshFromGateway].
     */
    override fun observeInstalled(): Flow<List<InstalledSkill>> = installedFlow.asStateFlow()

    /**
     * One-shot refresh from gateway `skills.list`.
     *
     * @return [Result] wrapping the refreshed list, or an exception on network failure.
     */
    override suspend fun refreshFromGateway(): Result<List<InstalledSkill>> = runCatching {
        val ws = wsFactory()
        val res = ws.request("skills.list", params = null).jsonObject
        val list = res["skills"]?.jsonArray.orEmpty().map { it.jsonObject.toInstalledSkill() }
        installedFlow.value = list
        list
    }

    /**
     * Starts an install via `skills.install`. The returned flow emits:
     * - [SkillInstallProgress.Connecting] immediately,
     * - zero or more intermediate progress notifications,
     * - a terminal [SkillInstallProgress.Succeeded] or [SkillInstallProgress.Failed].
     *
     * @param slug e.g. `"wanng-ide/auto-skill-hunter"`.
     * @param version `null` → latest.
     */
    @OptIn(FlowPreview::class)
    @Suppress("LongMethod")
    override fun install(slug: String, version: String?): Flow<SkillInstallProgress> = channelFlow {
        send(SkillInstallProgress.Connecting(slug))
        val ws = wsFactory()

        val notifJob = ws.notifications
            .filter { it.method == "skills.install.progress" }
            .onEach { n ->
                val p = n.params?.jsonObject ?: return@onEach
                val phase = p["phase"]?.jsonPrimitive?.contentOrNull
                val pct = p["pct"]?.jsonPrimitive?.intOrNull
                val progress: SkillInstallProgress? = when (phase) {
                    "downloading" -> SkillInstallProgress.Downloading(pct)
                    "verifying" -> SkillInstallProgress.Verifying(
                        sha256Prefix = p["sha256"]?.jsonPrimitive?.contentOrNull?.take(12) ?: ""
                    )
                    "applying" -> SkillInstallProgress.Applying(
                        step = p["step"]?.jsonPrimitive?.contentOrNull ?: "applying"
                    )
                    else -> null
                }
                if (progress != null) trySend(progress)
            }
            .launchIn(this)

        val params = buildJsonObject {
            put("source", JsonPrimitive("clawhub"))
            put("slug", JsonPrimitive(slug))
            if (version != null) put("version", JsonPrimitive(version))
            put("force", JsonPrimitive(false))
        }

        try {
            val result = ws.request("skills.install", params).jsonObject
            val skill = result.toInstalledSkill()
            installedFlow.value = (installedFlow.value.filter { it.slug != skill.slug } + skill)
            send(SkillInstallProgress.Succeeded(skill))
        } catch (e: GatewayWsClient.WsRpcException) {
            send(SkillInstallProgress.Failed(e.toInstallError()))
        } catch (e: Exception) {
            send(SkillInstallProgress.Failed(SkillInstallError.Unknown(e.message ?: "unknown")))
        } finally {
            notifJob.cancel()
        }
    }

    /**
     * Removes a previously-installed skill via `skills.uninstall`.
     *
     * @param slug The skill to remove.
     * @return [Result.success] on success, or a failure wrapping the exception.
     */
    override suspend fun uninstall(slug: String): Result<Unit> = runCatching {
        val ws = wsFactory()
        ws.request("skills.uninstall", buildJsonObject { put("slug", JsonPrimitive(slug)) })
        installedFlow.value = installedFlow.value.filter { it.slug != slug }
    }

    // ── helpers ───────────────────────────────────────────────────────────────

    private fun JsonObject.toInstalledSkill(): InstalledSkill = InstalledSkill(
        slug = this["slug"]!!.jsonPrimitive.content,
        installedVersion = this["installed_version"]!!.jsonPrimitive.content,
        source = when (this["source"]?.jsonPrimitive?.contentOrNull) {
            "ClawHub", "clawhub" -> SkillSource.ClawHub
            "Local", "local" -> SkillSource.Local
            else -> SkillSource.Unknown
        },
        installedAt = this["installed_at"]?.jsonPrimitive?.long ?: 0L,
    )

    private fun GatewayWsClient.WsRpcException.toInstallError(): SkillInstallError = when (code) {
        -32003 -> SkillInstallError.Unauthorized(missingScope = "operator.admin")
        -32004 -> SkillInstallError.NotFound(slug = message)
        -32005 -> SkillInstallError.DependencyConflict(conflicts = message.split(","))
        -32006 -> SkillInstallError.VerificationFailed(expected = "", actual = message)
        else -> SkillInstallError.Protocol(rpcCode = code, message = message)
    }
}

File: ios/GhostCrab/Data/Impl/InstalledSkillRepositoryImpl.swift
import Foundation

/// REST-backed ``InstalledSkillRepository``.
///
/// The Kotlin counterpart (`InstalledSkillRepositoryImpl.kt`) uses the
/// gateway's JSON-RPC WebSocket (`skills.list`, `skills.install`,
/// `skills.uninstall`) and streams intermediate progress notifications. The
/// iOS port talks to the equivalent REST endpoints on
/// ``OpenClawAPIClient`` instead — `getInstalledSkills()`,
/// `installSkill(slug:version:)`, `uninstallSkill(slug:)`. The REST install is
/// synchronous, so the install stream emits only:
///
/// - `.connecting(target:)` on subscribe, and
/// - a terminal `.succeeded(installed:)` or `.failed(error:)`.
///
/// The intermediate Kotlin phases (`downloading` / `verifying` / `applying`)
/// are unavailable over REST; the UI gracefully shows a single spinner
/// between connect and terminal.
///
/// State is held in a lock-protected box and republished to every observer via
/// an `AsyncStream` for each ``observeInstalled()`` call.
public final class InstalledSkillRepositoryImpl: InstalledSkillRepository, @unchecked Sendable {

    private let connectionManager: GatewayConnectionManagerImpl
    private let stateBox = InstalledSkillStateBox()

    /// - Parameter connectionManager: Concrete manager used to obtain the
    ///   active ``OpenClawAPIClient``.
    public init(connectionManager: GatewayConnectionManagerImpl) {
        self.connectionManager = connectionManager
    }

    /// Hot stream of the current installed-skill list. Starts empty; updated
    /// on successful install/uninstall and after ``refreshFromGateway()``.
    public func observeInstalled() -> AsyncStream<[InstalledSkill]> {
        stateBox.subscribe()
    }

    /// One-shot refresh from gateway `GET /api/skills`.
    public func refreshFromGateway() async -> Result<[InstalledSkill], any Error> {
        do {
            let client = try await connectionManager.requireClient()
            let dtos = try await client.getInstalledSkills()
            let list = dtos.map { $0.toInstalledSkill() }
            stateBox.set(list)
            return .success(list)
        } catch {
            return .failure(error)
        }
    }

    /// Start an install. The returned stream emits ``SkillInstallProgress``
    /// events and finishes after the terminal `.succeeded` / `.failed`.
    public func install(slug: String, version: String?) -> AsyncStream<SkillInstallProgress> {
        AsyncStream { continuation in
            continuation.yield(.connecting(target: slug))

            let task = Task { [weak self] in
                guard let self else {
                    continuation.finish()
                    return
                }
                do {
                    let client = try await self.connectionManager.requireClient()
                    let dto = try await client.installSkill(slug: slug, version: version)
                    let installed = dto.toInstalledSkill()
                    self.stateBox.upsert(installed)
                    continuation.yield(.succeeded(installed: installed))
                } catch let e as GatewayError {
                    continuation.yield(.failed(error: e.toSkillInstallError()))
                } catch is CancellationError {
                    // Caller cancelled the stream — no terminal event.
                } catch {
                    continuation.yield(.failed(error: .unknown(cause: error.localizedDescription)))
                }
                continuation.finish()
            }

            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }

    /// Remove a previously-installed skill via `POST /api/skills/uninstall`.
    public func uninstall(slug: String) async -> Result<Void, any Error> {
        do {
            let client = try await connectionManager.requireClient()
            try await client.uninstallSkill(slug: slug)
            stateBox.remove(slug: slug)
            return .success(())
        } catch {
            return .failure(error)
        }
    }
}

// ── State box ────────────────────────────────────────────────────────────────

/// Lock-protected mutable list of installed skills with multi-subscriber fan-out.
///
/// Replays the latest list on subscribe and yields every change to all live
/// subscribers — same semantics as Kotlin's `MutableStateFlow<List<InstalledSkill>>`.
private final class InstalledSkillStateBox: @unchecked Sendable {
    private var stored: [InstalledSkill] = []
    private var subscribers: [UUID: AsyncStream<[InstalledSkill]>.Continuation] = [:]
    private let lock = NSLock()

    func subscribe() -> AsyncStream<[InstalledSkill]> {
        AsyncStream { continuation in
            let id = UUID()
            self.lock.lock()
            let initial = self.stored
            self.subscribers[id] = continuation
            self.lock.unlock()

            continuation.yield(initial)

            continuation.onTermination = { [weak self] _ in
                guard let self else { return }
                self.lock.lock()
                self.subscribers.removeValue(forKey: id)
                self.lock.unlock()
            }
        }
    }

    func set(_ next: [InstalledSkill]) {
        lock.lock()
        stored = next
        let snapshot = next
        let subs = subscribers.values
        lock.unlock()
        for cont in subs {
            cont.yield(snapshot)
        }
    }

    func upsert(_ skill: InstalledSkill) {
        lock.lock()
        stored.removeAll { $0.slug == skill.slug }
        stored.append(skill)
        let snapshot = stored
        let subs = subscribers.values
        lock.unlock()
        for cont in subs {
            cont.yield(snapshot)
        }
    }

    func remove(slug: String) {
        lock.lock()
        stored.removeAll { $0.slug == slug }
        let snapshot = stored
        let subs = subscribers.values
        lock.unlock()
        for cont in subs {
            cont.yield(snapshot)
        }
    }
}

// ── Mapping ──────────────────────────────────────────────────────────────────

private extension InstalledSkillDTO {
    func toInstalledSkill() -> InstalledSkill {
        let resolvedSource: SkillSource
        switch source?.lowercased() {
        case "clawhub": resolvedSource = .clawHub
        case "local": resolvedSource = .local
        default: resolvedSource = .unknown
        }
        // Kotlin treats `installed_at` as epoch millis; default to 0 when absent.
        let resolvedDate: Date
        if let millis = self.installedAt {
            resolvedDate = Date(timeIntervalSince1970: TimeInterval(millis) / 1000)
        } else {
            resolvedDate = Date(timeIntervalSince1970: 0)
        }
        return InstalledSkill(
            slug: slug,
            installedVersion: installedVersion,
            source: resolvedSource,
            installedAt: resolvedDate
        )
    }
}

private extension GatewayError {
    /// Maps REST-level ``GatewayError`` cases to the install-time
    /// ``SkillInstallError`` taxonomy used by the install progress stream.
    ///
    /// Mirrors the Kotlin `WsRpcException.toInstallError()` mapping:
    /// - auth → `.unauthorized(missingScope: "operator.admin")`
    /// - api 404 → `.notFound(slug:)` (slug is unknown at this point; pass body)
    /// - api 409 → `.dependencyConflict(conflicts:)`
    /// - api 422 → `.verificationFailed(expected:actual:)`
    /// - api other → `.protocol(rpcCode:message:)`
    /// - timeout / unreachable → `.network(cause:)`
    /// - everything else → `.unknown(cause:)`
    func toSkillInstallError() -> SkillInstallError {
        switch self {
        case .auth:
            return .unauthorized(missingScope: "operator.admin")
        case let .api(_, statusCode, body):
            let bodyText = body ?? ""
            switch statusCode {
            case 404:
                return .notFound(slug: bodyText)
            case 409:
                return .dependencyConflict(conflicts: bodyText.split(separator: ",").map { String($0) })
            case 422:
                return .verificationFailed(expected: "", actual: bodyText)
            default:
                return .protocol(rpcCode: statusCode, message: bodyText)
            }
        case let .timeout(url, _):
            return .network(cause: "timeout at \(url)")
        case let .unreachable(url, _):
            return .network(cause: "unreachable at \(url)")
        case let .tls(url, _):
            return .network(cause: "TLS error at \(url)")
        case let .configValidation(field, reason):
            return .protocol(rpcCode: 0, message: "\(field): \(reason)")
        case .profileNeedsReauth:
            return .unauthorized(missingScope: "operator.admin")
        case let .aiServiceUnavailable(url):
            return .unknown(cause: "ai service unavailable at \(url)")
        case let .aiQuotaExceeded(url):
            return .unknown(cause: "ai quota exceeded at \(url)")
        }
    }
}
